Refactor brace completion to split into Features vs EditorFeatures

dibarbet opened the pull request:
Commit at a time is recommended, though the first commit is the bulk of the work :(

The first commit moves brace completion implementations from IEditorBraceCompletionSession to a new IBraceCompletionService that is available in the features layer (instead of editor features).  This allows us to use it in LSP.  
There were ~3 main issues with the existing impl
1.  All the IEditorBraceCompletionSession implementations [start off with editor concepts](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.EditorFeatures/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs,123) (text buffers, snapshots, IBraceCompletionSession) -  So I can't really use it in LSP.  Luckily in most of the cases (like the above) they just need some info that can be found from the document as well.  This enables nullable and makes things async where possible.
2.  The API shape for [IEditorBraceCompletionSession ](http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/AutomaticCompletion/IEditorBraceCompletionSession.cs,183d115f8b762132)isn't very LSP friendly -  the AfterStart and AfterReturn methods apply the changes to the workspace rather than giving back edits.  Additionally there's no way to return the actual brace completions or check if the caret is in a valid brace completion location.

The second commit uses the implementation to format braces after completion in razor via OnAutoInsert.  We will eventually want to use LSP for brace completion in liveshare/codespaces C# files (since we require semantics for <>), but since they don't yet support over-typing, regular .cs files will use the same local implementation that they do now (in liveshare and codespaces).

Additionally, we use OnAutoInsert instead of OnTypeFormatting since OnAutoInsert supports snippets (moving the caret location).  I previously spoke with @NTaylorMullen and the 'solution' for having OnTypeFormatting support snippets was OnAutoInsert.

@ajaybhargavb @NTaylorMullen there will be work on your side I believe to consume onautoinsert for this.

CyrusNajmabadi commented on Roslyn.sln:
```diff
@@ -1880,8 +1880,12 @@ Global
 		{9B25E472-DF94-4E24-9F5D-E487CE5A91FB}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{9B25E472-DF94-4E24-9F5D-E487CE5A91FB}.Release|Any CPU.Build.0 = Release|Any CPU
+		{9B25E472-DF94-4E24-9F5D-E487CE5A91FB}.Release|x86.ActiveCfg = Release|Any CPU
+		{9B25E472-DF94-4E24-9F5D-E487CE5A91FB}.Release|x86.Build.0 = Release|Any CPU
```
@jasonmalinowski everyone seems to be hitting this

jasonmalinowski replied:
@sharwell has a PR out to fix this.

sharwell replied:
Fix was merged

dibarbet commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,414 @@
+            return ImmutableArray.Create(BraceCompletionFormattingRule.ForIndentStyle(indentStyle)).AddRange(Formatter.GetDefaultFormattingRules(document));
+        }
+
+        private sealed class BraceCompletionFormattingRule : BaseFormattingRule
```
this type was just moved from the original curlybracecompletionsession

dibarbet commented:
@CyrusNajmabadi I think I'm ready for this to be looked at now - happy to hear your thoughts!

davidwengier reviewed (commented):
The LSP part of this makes sense to me (2nd commit). I don't feel too qualified to 
 comment on the first commit too much :P

dibarbet commented on src/EditorFeatures/CSharp/AutomaticCompletion/CSharpBraceCompletionSessionProvider.cs:
```diff
@@ -2,92 +2,31 @@
-            _undoManager = undoManager;
-        }
-
-        protected override bool IsSupportedOpeningBrace(char openingBrace)
```
code in here was moved to IBraceCompletionService.IsValidForBraceCompletion

dibarbet commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -69,34 +69,18 @@ public void Start()
                 var snapshot = SubjectBuffer.CurrentSnapshot;
                 var closingSnapshotPoint = ClosingPoint.GetPoint(snapshot);
 
-                if (closingSnapshotPoint.Position < 1)
```
The code to create the edit and caret location was moved to IBraceCompletionService.GetBraceCompletion
(implementation is in AbstractBraceCompletionService)

dibarbet commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/AbstractTokenBraceCompletionSession.cs:
```diff
@@ -1,88 +0,0 @@
-
-namespace Microsoft.CodeAnalysis.Editor.Implementation.AutomaticCompletion.Sessions
-{
-    internal abstract class AbstractTokenBraceCompletionSession : IEditorBraceCompletionSession
```
moved to AbstractTokenBraceCompletionService

dibarbet commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs:
```diff
@@ -0,0 +1,151 @@
+            _undoHistoryRegistry = undoHistoryRegistry;
+        }
+
+        public virtual void AfterReturn(IBraceCompletionSession session, CancellationToken cancellationToken)
```
this is code originally from CurlyBraceCompletionSession.AfterReturn, but heavily modified to generically apply any IBraceCompletionService.GetTextChangesAfterReturn

Currently the only implementation though is still CurlyBraceCompletionService

dibarbet commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,414 @@
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
+        }
+
+        public override async Task<BraceCompletionResult?> GetBraceCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
```
this code implements the CurlyBraceCompletionSession.AfterStart functionality (along with providing the actual closing brace edit)

dibarbet commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,414 @@
+            }
+        }
+
+        public override async Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext context, CancellationToken cancellationToken, bool supportsVirtualSpace = true)
```
this comes from CurlyBraceCompletionSession.AfterStart, with modifications to use documents and supporting non-virtual spaces for the caret location

dibarbet commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,414 @@
+            return Collapse(newDocumentText, changes.ToImmutableArray());
+        }
+
+        private static async Task<SourceText> FormatTrackingSpanAsync(Document document, int openingPoint, int closingPoint, bool shouldHonorAutoFormattingOnCloseBraceOption, CancellationToken cancellationToken, ImmutableArray<AbstractFormattingRule>? rules = null)
```
this is from the original method in CurlyBraceCompletionSession.FormatTrackingSpan, with extra logic added to apply the formatting to the document instead of editor subject buffer at the end.

dibarbet commented on src/Features/Core/Portable/BraceCompletion/AbstractTokenBraceCompletionService.cs:
```diff
@@ -0,0 +1,67 @@
+
+namespace Microsoft.CodeAnalysis.BraceCompletion
+{
+    internal abstract class AbstractTokenBraceCompletionService : AbstractBraceCompletionService
```
from abstracttokenbracecompletionsession

dibarbet commented on src/Features/Core/Portable/BraceCompletion/BraceCompletionOptions.cs:
```diff
@@ -0,0 +1,33 @@
+{
+    internal class BraceCompletionOptions
+    {
+        public static readonly PerLanguageOption2<bool> AutoFormattingOnCloseBrace = new(
```
Had to move these options down to the features layer

dibarbet commented on src/Features/VisualBasic/Portable/BraceCompletion/BracketBraceCompletionService.vb:
```diff
@@ -0,0 +1,90 @@
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+
+<Export(LanguageNames.VisualBasic, GetType(IBraceCompletionService)), [Shared]>
+Friend Class BracketBraceCompletionService
```
Pretty much all the VB implementation was just a straight move.

dibarbet commented on src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:
```diff
@@ -366,11 +366,11 @@ public Task<SemanticTokens> GetTextDocumentSemanticTokensRangeAsync(SemanticToke
 
         [JsonRpcMethod(MSLSPMethods.OnAutoInsertName, UseSingleObjectParameterDeserialization = true)]
-        public Task<DocumentOnAutoInsertResponseItem[]> GetDocumentOnAutoInsertAsync(DocumentOnAutoInsertParams autoInsertParams, CancellationToken cancellationToken)
+        public Task<DocumentOnAutoInsertResponseItem?> GetDocumentOnAutoInsertAsync(DocumentOnAutoInsertParams autoInsertParams, CancellationToken cancellationToken)
```
[] isn't valid - returning more than one fails

NTaylorMullen commented on src/Features/LanguageServer/Protocol/Handler/OnAutoInsert/OnAutoInsertHandler.cs:
```diff
@@ -58,20 +92,74 @@ public async Task<LSP.DocumentOnAutoInsertResponseItem[]> HandleRequestAsync(LSP
+            var position = sourceText.Lines.GetPosition(linePosition);
+
+            var serviceAndContext = await GetBraceCompletionContextAsync(position, document, cancellationToken).ConfigureAwait(false);
+            if (serviceAndContext != null)
```
nit: I always find it easier to consume by having multiple "if xyz => return null" to avoid all of the excess nesting which can be hard to visually parse.

```
if (serviceAndContext == null)
{
    return null;
}

.... The postReturnEdit check could also do this ....
```

NTaylorMullen commented on src/Features/LanguageServer/Protocol/Handler/OnAutoInsert/OnAutoInsertHandler.cs:
```diff
@@ -58,20 +92,74 @@ public async Task<LSP.DocumentOnAutoInsertResponseItem[]> HandleRequestAsync(LSP
+            if (serviceAndContext != null)
+            {
+                var (service, context) = serviceAndContext.Value;
+                var postReturnEdit = await service.GetTextChangeAfterReturnAsync(context, cancellationToken, supportsVirtualSpace: false).ConfigureAwait(false);
```
Prior to this I'd do a `cancellationToken.ThrowIfCancellationRequested()` since you already did some async work.

NTaylorMullen commented on src/Features/LanguageServer/Protocol/Handler/OnAutoInsert/OnAutoInsertHandler.cs:
```diff
@@ -3,48 +3,82 @@
 
             // The editor calls this handler for C# and VB comment characters, but we only need to process the one for the language that matches the document
-            if (autoInsertParams.Character != "\n" && autoInsertParams.Character != service.DocumentationCommentCharacter)
+            if (autoInsertParams.Character == "\n" || autoInsertParams.Character == service.DocumentationCommentCharacter)
```
Would be good to eventually move to a provider pattern, a lot like what we [have in Razor](https://github.com/dotnet/aspnetcore-tooling/blob/5a80d85b2cd38e0c63a25ca2f7f35fd55d7416c9/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/RazorOnAutoInsertProvider.cs) so these can all be easily expanded. Doesn't have to be in this PR but food for thought as you expand additional OnAutoInsert pieces (like @CyrusNajmabadi's comment newline support).

NTaylorMullen reviewed (commented):
👏

dibarbet commented:
sorry marking this as draft again - working on figuring out how to deal with tracking spans for the braces which the editor heavily relies on

dibarbet commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -364,6 +352,72 @@ private void MoveCaretToClosingPoint()
 
+            private void ApplyBraceCompletionResult(BraceCompletionResult result)
+            {
+                // Apply sets of edits incrementally and update the buffer on each set applied.
```
this is really unfortunate but I don't have a better solution here.  Open to ideas.

I did try just having the service specify the new open/close locations, but unfortunately the act of applying an edit triggers the check on the editor side for the tracking spans, and I cant change the tracking spans before applying the edit...

dibarbet commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -364,6 +352,72 @@ private void MoveCaretToClosingPoint()
+                    TextView.TryMoveCaretToAndEnsureVisible(snapshotPoint);
+                }
+
+                static void PutCaretOnLine(IBraceCompletionSession session, ISmartIndentationService smartIndentationService, int lineNumber)
```
all this was just moved from CurlyBraceCompletionSession

dibarbet commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,413 @@
+            return braceLocation;
+        }
+
+        private static async Task<ImmutableArray<TextChange>> FormatTrackingSpanAsync(Document document, int openingPoint, int closingPoint, bool shouldHonorAutoFormattingOnCloseBraceOption, CancellationToken cancellationToken, ImmutableArray<AbstractFormattingRule>? rules = null)
```
this method was just moved from the original CurlyBraceCompletionSession

dibarbet commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,413 @@
+
+        protected override bool IsValidClosingBraceToken(SyntaxToken token) => token.IsKind(SyntaxKind.CloseBraceToken);
+
+        private bool ContainsOnlyWhitespace(SourceText text, int openingPosition, int closingPosition)
```
this method was just moved from the original CurlyBraceCompletionSession

dibarbet commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,106 @@
+        /// Returns any text changes that need to be made after adding the closing brace.
+        /// </summary>
+        /// <remarks>
+        /// This cannot be merged with <see cref="GetBraceCompletionAsync(BraceCompletionContext, CancellationToken)"/>
```
this is also unfortunate, but again, I don't have any better ideas :(

CyrusNajmabadi replied:
seems ok tome :)

dibarbet commented:
@CyrusNajmabadi if you have time today, this could use a review - trying to get it for razor in p2

CyrusNajmabadi commented:
I won't.  But i can do tomorrow.

dibarbet commented:
> 
> 
> I won't. But i can do tomorrow.

works for me, thank you!

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractEditorBraceCompletionSessionFactory.cs:
```diff
@@ -2,48 +2,34 @@
-            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
-
-            return !syntaxFactsService.IsInNonUserCode(tree, position, cancellationToken);
+            return _braceCompletionSessions.SingleOrDefault(session => session.IsValidForBraceCompletion(openingBrace, openingPosition, document, cancellationToken));
```
SingleOrDefault is dangerous in hte case where two sessions say they're valid.  we'll throw in that case.  is that appropriate here?

dibarbet replied:
yes - there should only ever be one session, if there isn't it's a bug that we should know about.  I could nfw if preferred rather than throwing

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -94,50 +96,30 @@ private void Start(CancellationToken cancellationToken)
                 }
 
-                OpeningPoint = snapshot.CreateTrackingPoint(openingSnapshotPoint, PointTrackingMode.Positive);
+                OpeningPoint = SubjectBuffer.CurrentSnapshot.CreateTrackingPoint(openingSnapshotPoint, PointTrackingMode.Positive);
```
since we are accessing .CurrentSnapshot, it would be a good idea for this type to either derive from, or contain a ForegroundThreadAffinitizedObject.  We woudl then assert in these methods htat were were called on the UI thread.  thus expressing hte necessary constraints for our caller, and throwing immediately if that is ever violated.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -94,50 +96,30 @@ private void Start(CancellationToken cancellationToken)
+                // Apply the change to complete the brace.
+                ApplyBraceCompletionResult(braceResult.Value);
 
                 // switch from positive to negative tracking so it stays against the closing brace
```
what is "it" in "so it stays"?

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -268,7 +250,13 @@ public void PostReturn()
                     if (closingSnapshotPoint.Position > 0 && HasNoForwardTyping(this.GetCaretPosition().Value, closingSnapshotPoint.Subtract(1)))
                     {
-                        _session.AfterReturn(this, CancellationToken.None);
+                        var changesAfterReturn = _session.GetChangesAfterReturn(this, CancellationToken.None);
```
`CancellationToken.None`

:(

is thee no cancellation token, or OperationContext that is passed in tha twe can use here?

dibarbet replied:
there is nothing passed to us here from the editor :(

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -364,6 +352,61 @@ private void MoveCaretToClosingPoint()
                 }
             }
 
+            private void ApplyBraceCompletionResult(BraceCompletionResult result)
```
is ths called from multiple places?  if not, should we move the transaction *into* this?

dibarbet replied:
this is called from multiple places.  In the first two instances it is used to insert the closing brace result, then apply the formatting changes result all in one transaction.  In the second instance it inserts the changes after return in a transaction.  So the transaction needs to be outside of this method.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -364,6 +352,61 @@ private void MoveCaretToClosingPoint()
+                Debug.Assert(SubjectBuffer.CurrentSnapshot[OpeningPoint.GetPosition(SubjectBuffer.CurrentSnapshot)] == OpeningBrace,
+                "The opening point does not match the opening brace character");
+                Debug.Assert(SubjectBuffer.CurrentSnapshot[ClosingPoint.GetPosition(SubjectBuffer.CurrentSnapshot) - 1] == ClosingBrace,
+                "The closing point does not match the closing brace character");
```
consider making these NFWs.  we'll never hear about debug-asserts as no one runs in debug.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -364,6 +352,61 @@ private void MoveCaretToClosingPoint()
+                if (line.GetText().IsNullOrWhiteSpace())
+                {
+                    // If the caret is on an empty line, put it at the correct indentation using virtual space.
+                    PutCaretOnLine(this, _smartIndentationService, line.LineNumber);
```
nit: it seems unfortunate to me that this value is not part of the BraceCompletionResult .  i would think *any* host would want this info to set the caret properly, and it should just be part of hte 'data' we compute and return.

dibarbet replied:
The caret location is returned - I can add an additional property for the indentation if on a new line

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -364,6 +352,61 @@ private void MoveCaretToClosingPoint()
+                {
+                    // Not on an empty line, just put the caret at the specified location.
+                    TextView.TryMoveCaretToAndEnsureVisible(snapshotPoint);
+                }
```
```suggestion
                }
                
                return;
```

placing an explicit return before the local functions makes it immediately obvious that hte main 'code' of hte method is done and the locals are not interspersed with it.  without this, one has to scan all the way to the end to see if there's any more interesting code that is executed.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -364,6 +352,61 @@ private void MoveCaretToClosingPoint()
+                        return indentation.Value;
+                    }
+
+                    // do poor man's indentation
```
do you ever hit this?

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/IEditorBraceCompletionSession.cs:
```diff
@@ -2,19 +2,19 @@
+        BraceCompletionResult? GetChangesAfterCompletion(IBraceCompletionSession session, CancellationToken cancellationToken);
         bool AllowOverType(IBraceCompletionSession session, CancellationToken cancellationToken);
-        void AfterReturn(IBraceCompletionSession session, CancellationToken cancellationToken);
+        BraceCompletionResult? GetChangesAfterReturn(IBraceCompletionSession session, CancellationToken cancellationToken);
```
group these method better (i.e. put the GetChangesXXX methods together).  Please doc the distinction and semantics as this is not obvious.  i.e. what's the diff from GetBraceCompletion and GetChangesAfterCompletion.  why do both return the same type?  (i honestly don't know/understand).

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs:
```diff
@@ -0,0 +1,84 @@
+                return null;
+            }
+
+            return _braceCompletionService.GetTextChangeAfterReturnAsync(context, supportsVirtualSpace: true, cancellationToken).WaitAndGetResult(cancellationToken);
```
in general, asynchrony should not be blocked at the leaves.  instead push the asynchrony up *as far as you can* and do the waiting explicitly at the highest level possible.  This keeps all the services/helpers/etc. nice and clean, and keeps the nasty unfortunate stuff isolated to one place.

dibarbet replied:
Since this type is basically now only used to call the async brace completion service from the sync BraceCompletionSession, I'll just move these calls up to BraceCompletionSession and delete this type

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs:
```diff
@@ -0,0 +1,84 @@
+            return _braceCompletionService.GetTextChangeAfterReturnAsync(context, supportsVirtualSpace: true, cancellationToken).WaitAndGetResult(cancellationToken);
+        }
+
+        public virtual bool AllowOverType(IBraceCompletionSession session, CancellationToken cancellationToken)
```
don't love that we have virtual behavior in a concrete type.  will have to see why we do this and if we should change.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs:
```diff
@@ -0,0 +1,84 @@
+        public virtual bool AllowOverType(IBraceCompletionSession session, CancellationToken cancellationToken)
+        {
+            var context = GetBraceCompletionContextFromSession(session);
+            if (context != null)
+            {
+                return _braceCompletionService.AllowOverTypeAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
+            }
```
`return context == null ||`

Also, please doc why we return 'true' in the 'null' case, it's not self evident.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs:
```diff
@@ -0,0 +1,84 @@
+        }
+
+        public bool IsValidForBraceCompletion(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+            => _braceCompletionService.IsValidForBraceCompletionAsync(brace, openingPosition, document, cancellationToken).WaitAndGetResult(cancellationToken);
```
reminder about pushign async/wait upwards.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs:
```diff
@@ -0,0 +1,84 @@
+                return new BraceCompletionContext(document, openingSnapshotPoint, closingSnapshotPoint, caretPosition);
+            }
+
+            return null;
```
in general, please flip these checks.  in other words don't do:

```
if (!something)
{
     big complex logic
}

// bail out logic
```

do

```
if (something)
   bail out

// big complex logic
```

this is much easier to grok as the simple case can be understood and validated immediately, before finally getting to the 'meat'.  with the above form, one has to examine the complex case, then go back and check the condition again to see if it fits the bail out case.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs:
```diff
@@ -0,0 +1,84 @@
+
+        private static BraceCompletionContext? GetBraceCompletionContextFromSession(IBraceCompletionSession session)
+        {
+            var snapshot = session.SubjectBuffer.CurrentSnapshot;
```
1. this type also would benefit from threading guarantees
2. instead of having this type grab from the subject buffer (something inherently mutable), it would be better if the caller grabbed hte immutable pieces that were needed and passed them into this type

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs:
```diff
@@ -0,0 +1,84 @@
+            {
+                var closingSnapshotPoint = session.ClosingPoint.GetPosition(snapshot);
+                var openingSnapshotPoint = session.OpeningPoint.GetPosition(snapshot);
+                var caretPosition = session.GetCaretPosition()?.Position;
```
doc this.  what does it mean for the caret to be null??

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/BracketBraceCompletionService.cs:
```diff
@@ -0,0 +1,37 @@
+        public override async Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            return await CheckCurrentPositionAsync(context.Document, context.CaretLocation, cancellationToken).ConfigureAwait(false)
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
```
so this logic isn't apperant to me.  why is overtype ok if the currentposition/closingtoken is ok?

for example, if the above code were just `CheckCurrentPositionAsync` or `CheckClosingTokenKindAsync` i would have no idea if that were right/wrong.

dibarbet replied:
I'm resolving this by moving these checks to a method (with much nicer name) in the base and doc'ing that method.
I also unified a bunch of the AllowOverType implementations which should make everything clearer (and resolve the other allowovertypeasync comments)

CyrusNajmabadi replied:
excellent!

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CharLiteralBraceCompletionService.cs:
```diff
@@ -0,0 +1,35 @@
+        protected override char ClosingBrace => SingleQuote.CloseCharacter;
+
+        public override Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.True;
```
doc this.  i think i can understand why it is the cae, but it's not totally clear (esp. for a future reader).

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+    [Export(LanguageNames.CSharp, typeof(IBraceCompletionService)), Shared]
+    internal class CurlyBraceCompletionService : AbstractBraceCompletionService
+    {
+        private static readonly SyntaxAnnotation s_closingBraceSyntaxAnnotation = new("original closing brace");
```
doc what this id for.

CyrusNajmabadi replied:
also, why do we need this for culies, but not for BracketBraceCompletionService ?

dibarbet replied:
bracket brace completion doesn't do anything special afterstart / afterreturn so there isn't complexity in finding the closing brace.  But for curlies we do formatting after start and after return.  Previously it used the closing point tracking span to do find the closing brace location (to put the caret there) but those aren't available in this layer

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+        public override async Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            return await CheckCurrentPositionAsync(context.Document, context.CaretLocation, cancellationToken).ConfigureAwait(false)
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
```
shared logic with another service.  should tehre be a well-named helper that expresses this idea that we can doc and which can be used from all the similar services?

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+
+        public override async Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+        {
+            // Format the span from the open brace location up to and including the closing brace location.
```
this is a what comment, not a why comment.  can you add the 'why', perhaps with an example?  i can guess, but comments take the guesswork out :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+        public override async Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+        {
+            // Format the span from the open brace location up to and including the closing brace location.
+            var (formattingChanges, newClosingPoint) = await FormatTrackingSpanAsync(braceCompletionContext.Document, braceCompletionContext.OpeningPoint, braceCompletionContext.ClosingPoint,
+                shouldHonorAutoFormattingOnCloseBraceOption: true, rules: null, cancellationToken).ConfigureAwait(false);
```
```suggestion
            var (formattingChanges, newClosingPoint) = await FormatTrackingSpanAsync(
                braceCompletionContext.Document, braceCompletionContext.OpeningPoint, braceCompletionContext.ClosingPoint,
                shouldHonorAutoFormattingOnCloseBraceOption: true, rules: null, cancellationToken).ConfigureAwait(false);
```

CyrusNajmabadi replied:
can you doc why no rules?

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+            }
+
+            // The caret location should be at the start of the closing brace character.
+            return new BraceCompletionResult(formattingChanges, newClosingPoint - 1);
```
so `newClosingPoint` is *after* the close brace?  can you call is finalCurlyBraceEnd then, or something like that?

dibarbet replied:
yeah the closing point of the session is the location immediately following the brace, I tried to keep that consistent.  I'll update the name

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+            var documentSnapshotText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            // Create a new source text that is not based on an editor snapshot so we can take advantage
+            // of the multi-version text change merging in ChangedText instead of ChangedSnapshotText.
+            var originalDocumentText = SourceText.From(documentSnapshotText.ToString(), documentSnapshotText.Encoding, documentSnapshotText.ChecksumAlgorithm);
```
=-o...

i am... really not loving this.  esp as that .ToString and SourceText.From could be very expensive.

CyrusNajmabadi replied:
i presume you're using this to get access to the 'Merge' helper in that type.  But it feels to me like that could be extracted out and we could operate on the low level pieces of data (i.e. the TextChangesRanges) directly, accessing tha thelper to get the final result.  is that not hte case?

dibarbet replied:
Yeah essentially - I can do that, but if you don't mind I'll do that after the rest of the PR looks good as I'll need a compiler review on the change in ChangedText.  I'd rather not spam them too much yet :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+            }
+            else
+            {
+                // The caller does not support virtual spaces, so we have to insert actual whitespace into the document.
```
fwiw, i think this should potentially be hte callers responsibility.  but i'm ont going to push hard on that if it turns out to be extremely difficult to do.

but it does seem like this API should just return data.  So it woudl say: here are the text changes.  And after that text change, the caret should be on line 'X' at positoin 'Y'.  Host then says: sure, i'll place it htere with virtual space.  Or "can't do that.  i'll insert whitespcae".

The concept of 'virtual space' should not leak into 'Features'.

CyrusNajmabadi replied:
this also feels like this solve the "i need to pass back (or merge) multiple set of textchangeranges"

this helper will always just give you the formatting changes, and the final desired line/col.  The higher up layer will be responsible for handling that.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+            return true;
+
+            static int GetValueInValidRange(int value, int smallest, int largest)
+            => Math.Max(smallest, Math.Min(value, largest));
```
indent.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+            start = text[start] == OpeningBrace ? start + 1 : start;
+
+            var end = closingPosition - 1;
+            end = text[end] == ClosingBrace ? end - 1 : end;
```
this is so weird, esp. as you might have multiple of a closebrace next to each other.  i honestly don't get this code.

dibarbet replied:
Yep I didn't get this one either.  It's a straight copy but I'll see if I can clean it up

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+            end = text[end] == ClosingBrace ? end - 1 : end;
+
+            if (!PositionInSnapshot(start, text) ||
+                !PositionInSnapshot(end, text))
```
wtf...

CyrusNajmabadi replied:
even if this is a copy, this function is awful :-/

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+        /// Returns the text changes that should be applied to the input document to 
+        /// get the formatted text and the new closing point in the formatted text.
+        /// </summary>
+        private static async Task<(ImmutableArray<TextChange> TextChanges, int NewClosingPoint)> FormatTrackingSpanAsync(
```
```suggestion
        private static async Task<(ImmutableArray<TextChange> textChanges, int newClosingPoint)> FormatTrackingSpanAsync(
```

we do not capitalize tuple members.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+                        endPoint -= endToken.Span.Length + startToken.Span.Length;
+                    }
+                }
+            }
```
i don't really understand thse comments.  the first comment says "do not format wihin the braces.  but hte second comment says "format everything but the braces...

i don't get it.  nor do i even understand the logic.  if we're formatting eveyrthing but hte brace pair... shouldn't the start point move forward, and the endpoint backward?

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+                var root = originalRoot.ReplaceToken(closeBraceToken, newCloseBraceToken);
+                return document.WithSyntaxRoot(root);
+            }
+        }
```
i gave up on this method.  needless to say, i hate it.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+            }
+        }
+
+        private static ImmutableArray<AbstractFormattingRule> GetBraceFormattingRules(Document document)
```
why is this method not async.

CyrusNajmabadi replied:
it's called from an async locatin, but then hard waits :-/

CyrusNajmabadi replied:
the caller also grabs the options themslves.  this is bizarre.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+                (currentToken.Parent.Kind() == SyntaxKind.ObjectInitializerExpression ||
+                currentToken.Parent.Kind() == SyntaxKind.CollectionInitializerExpression ||
+                currentToken.Parent.Kind() == SyntaxKind.ArrayInitializerExpression ||
+                currentToken.Parent.Kind() == SyntaxKind.ImplicitArrayCreationExpression))
```
indentation.

CyrusNajmabadi replied:
use .IsParentKind, then ou don't need the null check.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+                // new List<int> {
+                // int[] arr = {
+                //           = new[] {
+                //           = new int[] {
```
this comment doesn't help me at all.  so... it's these cases.  then what?

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+                        AddAlignIndentationOfTokensToBaseTokenOperation(list, node, bracePair.Item1, SpecializedCollections.SingletonEnumerable(bracePair.Item2), AlignTokensOption.AlignIndentationOfTokensToFirstTokenOfBaseTokenLine);
+                    }
+                }
+            }
```
needs docs.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,416 @@
+                {
+                    // remove any suppression operation
+                    list.RemoveAll(s_predicate);
+                }
```
needs docs.  this says what it is doing, but not why.  i have no idea how to assess this for correctness.

dibarbet replied:
Not sure to be honest either.  It looks like it's removing suppression rules that suppress wrapping, but the formatter is magic to me.  If I remove this a bunch of tests fail with the wrong indentation.  Put a comment marking as such and moving on.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/InterpolatedStringBraceCompletionService.cs:
```diff
@@ -0,0 +1,84 @@
+            => SpecializedTasks.True;
+
+        public override async Task<bool> IsValidForBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+            => OpeningBrace == brace && await IsPositionInInterpolatedStringContextAsync(document, openingPosition, cancellationToken).ConfigureAwait(false);
```
needs docs.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/InterpolatedStringBraceCompletionService.cs:
```diff
@@ -0,0 +1,84 @@
+        protected override char ClosingBrace => DoubleQuote.CloseCharacter;
+
+        public override Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+            => SpecializedTasks.True;
```
needs docs.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/InterpolatedStringBraceCompletionService.cs:
```diff
@@ -0,0 +1,84 @@
+
+            return root.SyntaxTree.IsExpressionContext(start, token, attributes: false, cancellationToken: cancellationToken)
+                || root.SyntaxTree.IsStatementContext(start, token, cancellationToken);
+        }
```
entire type needs docs.  it's super unclear.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/InterpolationBraceCompletionService.cs:
```diff
@@ -0,0 +1,97 @@
+                SyntaxKind.InterpolatedStringEndToken);
+        }
+    }
+}
```
same comment here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/LessAndGreaterThanBraceCompletionService.cs:
```diff
@@ -2,51 +2,55 @@
                 !token.CheckParent<TypeArgumentListSyntax>(n => n.LessThanToken == token) &&
                 !token.CheckParent<FunctionPointerParameterListSyntax>(n => n.LessThanToken == token) &&
-                !PossibleTypeArgument(snapshot, token, cancellationToken))
+                !await PossibleTypeArgumentAsync(document, token, cancellationToken).ConfigureAwait(false))
```
this is strange.  only if *all* of these are not true, we say it's not valid?

dibarbet replied:
Not sure on the question here, the check makes sense to me I think.  If the < token is not for any kind of type argument it is not valid for brace completion (since we don't want to complete a < conditional).

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/ParenthesisBraceCompletionService.cs:
```diff
@@ -0,0 +1,65 @@
+        public override async Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            return await CheckCurrentPositionAsync(context.Document, context.CaretLocation, cancellationToken).ConfigureAwait(false)
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
```
another repeat.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/ParenthesisBraceCompletionService.cs:
```diff
@@ -0,0 +1,65 @@
+        protected override async Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            // check token at the opening point first
```
this is a what, not a why.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/ParenthesisBraceCompletionService.cs:
```diff
@@ -0,0 +1,65 @@
+            // now check whether parser think whether there is already counterpart closing parenthesis
+            var (openBrace, closeBrace) = token.Parent.GetParentheses();
+
+            // if pair is on the same line, then the closing parenthesis must belong to other tracker.
```
i don't know what this means.  what "other tracker"?

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/StringLiteralBraceCompletionService.cs:
```diff
@@ -0,0 +1,64 @@
+        protected override char ClosingBrace => DoubleQuote.CloseCharacter;
+
+        public override Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+            => SpecializedTasks.True;
```
doc.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/StringLiteralBraceCompletionService.cs:
```diff
@@ -0,0 +1,64 @@
+            }
+
+            return Task.FromResult(token.SpanStart + 1 == position && token.IsVerbatimStringLiteral());
+        }
```
doc.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,178 @@
+        /// </summary>
+        protected abstract bool IsValidClosingBraceToken(SyntaxToken token);
+
+        ///<inheritdoc cref="IBraceCompletionService.AllowOverTypeAsync(BraceCompletionContext, CancellationToken)"/>
+        public abstract Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);
+
+        ///<inheritdoc cref="IBraceCompletionService.GetBraceCompletionAsync(BraceCompletionContext, CancellationToken)"/>
+        public async Task<BraceCompletionResult?> GetBraceCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
```
```suggestion
        /// <inheritdoc cref="IBraceCompletionService.AllowOverTypeAsync(BraceCompletionContext, CancellationToken)"/>
        public abstract Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);

        /// <inheritdoc cref="IBraceCompletionService.GetBraceCompletionAsync(BraceCompletionContext, CancellationToken)"/>
        public async Task<BraceCompletionResult?> GetBraceCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
```

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,178 @@
+
+            var braceTextEdit = new TextChange(TextSpan.FromBounds(closingPoint, closingPoint), ClosingBrace.ToString());
+            // The caret location should be right before where the closing brace was inserted.
+            return new BraceCompletionResult(ImmutableArray.Create(braceTextEdit), caretLocation: openingPoint + 1);
```
codedoesn't match comment.  comment talks aobut closing-brace, but code sets the caret based on open brace.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,178 @@
+            return new BraceCompletionResult(ImmutableArray.Create(braceTextEdit), caretLocation: openingPoint + 1);
+        }
+
+        ///<inheritdoc cref="IBraceCompletionService.GetTextChangesAfterCompletionAsync(BraceCompletionContext, CancellationToken)"/>
+        public virtual Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        ///<inheritdoc cref="IBraceCompletionService.GetTextChangeAfterReturnAsync(BraceCompletionContext, bool, CancellationToken)"/>
+        public virtual Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext braceCompletionContext, bool supportsVirtualSpace, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        ///<inheritdoc cref="IBraceCompletionService.IsValidForBraceCompletionAsync(char, int, Document, CancellationToken)"/>
+        public virtual async Task<bool> IsValidForBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
```
```suggestion
        /// <inheritdoc cref="IBraceCompletionService.GetTextChangesAfterCompletionAsync(BraceCompletionContext, CancellationToken)"/>
        public virtual Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
            => SpecializedTasks.Default<BraceCompletionResult?>();

        /// <inheritdoc cref="IBraceCompletionService.GetTextChangeAfterReturnAsync(BraceCompletionContext, bool, CancellationToken)"/>
        public virtual Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext braceCompletionContext, bool supportsVirtualSpace, CancellationToken cancellationToken)
            => SpecializedTasks.Default<BraceCompletionResult?>();

        /// <inheritdoc cref="IBraceCompletionService.IsValidForBraceCompletionAsync(char, int, Document, CancellationToken)"/>
        public virtual async Task<bool> IsValidForBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
```

CyrusNajmabadi replied:
fwiw, i don't think you need these inherit-docs.  if these are interface impls, don't we get this for free?

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,178 @@
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return OpeningBrace == brace && !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
```
can you check the OpeningBrace thing first, then actually do the exensive stuff like getting the tree?

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,178 @@
+        protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            if (!IsParentSkippedTokensTrivia(syntaxFactsService, token))
```
doc.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,178 @@
+                return SpecializedTasks.False;
+            }
+
+            return Task.FromResult(IsValidOpeningBraceToken(token) && token.SpanStart == position);
```
can you do the cheap check first?

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,178 @@
+        protected static async Task<bool> CheckCurrentPositionAsync(Document document, int? currentPosition, CancellationToken cancellationToken)
+        {
+            // make sure auto closing is called from a valid position
+            if (currentPosition == null)
```
this feels weird.  who is calling this with a possibly null position?  wouldn't they just bail out?  when is a null positoin ok to proceed?  when does this happen?

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,178 @@
+        }
+
+        protected static bool IsParentSkippedTokensTrivia(ISyntaxFactsService syntaxFactsService, SyntaxToken token)
+            => token.Parent != null && !syntaxFactsService.IsSkippedTokensTrivia(token.Parent);
```
i don't see how this possibly works.  the method is called IsParentSkippedTokensTrivia, but then you return ` !syntaxFactsService.IsSkippedTokensTrivia(token.Parent);`  

this also means we're missing a test afaict.

dibarbet replied:
the usages of the check are correct, I just renamed this wrong - I'm going to switch the impl to be the name as it makes more sense to me to do the check that way

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,94 @@
+        /// Note that this does not mean that the service will provide brace completion, but that
+        /// only this service will be asked to provide them.
+        /// </summary>
+        Task<bool> IsValidForBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken);
```
CanProvideBraceCompletion.

CyrusNajmabadi replied:
can you doc the args?  from looking at it, it seems redundant.  i.e. Doc+OpenPosition = brace.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,94 @@
+        /// <remarks>
+        /// This cannot be merged with <see cref="GetBraceCompletionAsync(BraceCompletionContext, CancellationToken)"/>
+        /// as we need to make modifications to tracking spans in between adding the closing brace and doing
+        /// any kind of formatting on it.  So these must be two distinct steps until we fully move to LSP.
```
i don't really understand the explanatoin.  can you provide an example?

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,94 @@
+        /// Returns the brace completion context if the caret is located between an already completed
+        /// set of braces with only whitespace in between.
+        /// </summary>
+        Task<BraceCompletionContext?> IsInsideCompletedBracesAsync(int caretLocation, Document document, CancellationToken cancellationToken);
```
calling this 'Is', but not returning a boolean is pretty weird.

also, this should be `Doc, location`, not `location, Doc`

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,94 @@
+
+        /// <summary>
+        /// Whether or not overtype should be allowed given a brace completion context.
+        /// </summary>
```
doc this more.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,94 @@
+        Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);
+    }
+
+    internal struct BraceCompletionResult
```
readonly

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    internal class BraceCompletionContext
```
why a class, and not a readonly struct?

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,94 @@
+
+        public int ClosingPoint { get; }
+
+        public int? CaretLocation { get; }
```
doc all.  also, what does it mean to have a null caret location?

CyrusNajmabadi commented on src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs:
```diff
@@ -154,5 +155,14 @@ public static async Task<NamingRule> GetApplicableNamingRuleAsync(this Document
+        {
+            var workspace = document.Project.Solution.Workspace;
+            var formattingRuleFactory = workspace.Services.GetRequiredService<IHostDependentFormattingRuleFactoryService>();
+            var position = (span.Start + span.End) / 2;
```
doc..

dibarbet replied:
I dug into this, it looks like it's venus related but honestly I have no idea why the midpoint is better than the start or end (or why it doesn't take in a span?  there's no docs on IHostDependentFormattingRuleFactoryService either...)

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/OnAutoInsert/OnAutoInsertHandler.cs:
```diff
@@ -3,48 +3,85 @@
 
             // The editor calls this handler for C# and VB comment characters, but we only need to process the one for the language that matches the document
-            if (autoInsertParams.Character != "\n" && autoInsertParams.Character != service.DocumentationCommentCharacter)
+            if (autoInsertParams.Character == "\n" || autoInsertParams.Character == service.DocumentationCommentCharacter)
```
i honestly don't understand this check.

dibarbet replied:
This is just checking that the character the LSP client called into for onautoinsert is one that should trigger the documentation comment service.  It's triggered either on a new line or the whatever character the documentation comment service says it should be (e.g. in c#, `/` or in VB `'`)

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/OnAutoInsert/OnAutoInsertHandler.cs:
```diff
@@ -58,20 +95,83 @@ public async Task<LSP.DocumentOnAutoInsertResponseItem[]> HandleRequestAsync(LSP
+            }
+
+            var (service, context) = serviceAndContext.Value;
+            var postReturnEdit = await service.GetTextChangeAfterReturnAsync(context, supportsVirtualSpace: false, cancellationToken).ConfigureAwait(false);
```
here is where we'd get the result back, then determine the whitespace to insert due to virtual space.

CyrusNajmabadi commented on src/Features/LanguageServer/ProtocolUnitTests/OnAutoInsert/OnAutoInsertTests.cs:
```diff
@@ -179,16 +177,79 @@ string M(int foo, bool bar)
+    {
+        if (true)
+        {
+            $0
```
this is strange.  what if there is a C#/Vb string with this text in it?

dibarbet replied:
LSP spec on snippets - https://microsoft.github.io/language-server-protocol/specification#snippet_syntax

If there is a snippet containing both a tab stop($0) at some location and the text '$0' in a string somewhere I think we're basically hosed...

For non-snippets we're generally fine as we specify the plain text format and the client won't try to interpret the $0 as caret locations.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/BraceCompletion/CurlyBraceCompletionService.vb:
```diff
@@ -0,0 +1,54 @@
+        MyBase.New()
+    End Sub
+
+    Protected Overrides ReadOnly Property OpeningBrace As Char
+        Get
+            Return CurlyBrace.OpenCharacter
+        End Get
+    End Property
```
```suggestion
    Protected Overrides ReadOnly Property OpeningBrace As Char = CurlyBrace.OpenCharacter
```

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/BraceCompletion/CurlyBraceCompletionService.vb:
```diff
@@ -0,0 +1,54 @@
+
+    Public Overrides Async Function AllowOverTypeAsync(context As BraceCompletionContext, cancellationToken As CancellationToken) As Task(Of Boolean)
+        Return Await CheckCurrentPositionAsync(context.Document, context.CaretLocation, cancellationToken).ConfigureAwait(False) _
+            And Await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(False)
```
shared helper.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/BraceCompletion/InterpolatedStringBraceCompletionService.vb:
```diff
@@ -0,0 +1,64 @@
+        Dim text = Await document.GetTextAsync(cancellationToken).ConfigureAwait(False)
+
+        Return text(position - 1) = "$"c
+    End Function
```
am i wrong, or is the Vb side of this very different from the c# side.  if so, why?

dibarbet replied:
The VB side is very different from the C# side.  I don't entirely know the reasons so I did my best to just move the code and not change the logic inthis PR

dibarbet replied:
Dug into this one for a bit, the corresponding methods used by C# are either not available for VB (IsStatementContext) or do not return the expected results (IsExpressionContext) for VB.  It doesn't seem to think it's in an expression if you do `Dim s = $`.  Going to leave this as-is as I don't have a better idea of a check here.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/BraceCompletion/InterpolationBraceCompletionService.vb:
```diff
@@ -0,0 +1,89 @@
+
+        If openCurlyCount Mod 2 > 0 Then
+            Return False
+        End If
```
=-o.  wat

dibarbet replied:
this is similar to the logic in the c# version... unifying

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/BraceCompletion/ParenthesisBraceCompletionService.vb:
```diff
@@ -0,0 +1,63 @@
+        Dim skippedTriviaNode = TryCast(token.Parent, SkippedTokensTriviaSyntax)
+        If skippedTriviaNode IsNot Nothing Then
+            Dim skippedToken = skippedTriviaNode.ParentTrivia.Token
+            If skippedToken.Kind <> SyntaxKind.CloseParenToken OrElse Not TypeOf skippedToken.Parent Is BinaryConditionalExpressionSyntax Then
```
wat?

i honestly have no undestanding of any of tehse checs.  why is BinaryConditional special cased?  what about the ternary conditional form?  or other forms that end with a close paren??

dibarbet replied:
I spent half an hour playing with this tonight and couldn't come up with any scenario where the checks here made sense.  Going to put a comment and leave as-is because I'm thoroughly confused (and git history is not helpful)

CyrusNajmabadi commented:
This entire area needs a lot of TLC.

dibarbet commented:
@CyrusNajmabadi I think this is ready for round 2 - it will probably be useful to review commit by commit as I've split them out by the feedback they're addressing.

Note I haven't addressed the comment about Source text merging (still need it as I apply new line, then formatting changes which need merging).  My plan is to make those changes once everything else is signed off to avoid spamming the compiler team with lots of comments :)

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -76,6 +77,7 @@ public void Start()
 
             private void Start(CancellationToken cancellationToken)
             {
+                this.AssertIsForeground();
```
awesome.  this is def much easier ot understand for me.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -371,10 +373,18 @@ private void ApplyBraceCompletionResult(BraceCompletionResult result)
+                    return;
+                }
+                
 
```
two newlines.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -385,45 +383,8 @@ private void ApplyBraceCompletionResult(BraceCompletionResult result)
-                    return openingPoint - openingSpanLine.Start;
-                }
+                var caretLine = SubjectBuffer.CurrentSnapshot.GetLineFromLineNumber(result.CaretLocation.Line);
+                TextView.TryMoveCaretToAndEnsureVisible(new VirtualSnapshotPoint(caretLine, result.CaretLocation.Character));
```
love it.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs:
```diff
@@ -27,7 +27,7 @@ public EditorBraceCompletionSession(IBraceCompletionService braceCompletionServi
             }
 
-            return _braceCompletionService.GetTextChangeAfterReturnAsync(context, supportsVirtualSpace: true, cancellationToken).WaitAndGetResult(cancellationToken);
+            return _braceCompletionService.GetTextChangeAfterReturnAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
```
👍

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -101,40 +105,30 @@ public override async Task<bool> AllowOverTypeAsync(BraceCompletionContext conte
+                var offsetOfBacePosition = baseLinePosition.Character;
+                var totalOffset = offsetOfBacePosition + indentation.Offset;
+                var indentedLinePosition = new LinePosition(lineNumber, totalOffset);
+                return indentedLinePosition;
```
not related to this code, just a general question:

Do our tests somehow validate the indentation position of brace completion?

dibarbet replied:
They do, both the caret and text
http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.Test.Utilities/AutomaticCompletion/AbstractAutomaticBraceCompletionTests.cs,86

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -64,9 +64,9 @@ internal struct BraceCompletionResult
         /// to the input text.
         /// </summary>
-        public int CaretLocation { get; }
+        public LinePosition CaretLocation { get; }
```
might be worth pointing out that this can be 'virtual' (or however you want to explain that it might refer to a desired column that doesn't point at actual whitespace.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/OnAutoInsert/OnAutoInsertHandler.cs:
```diff
@@ -122,17 +123,34 @@ internal class OnAutoInsertHandler : IRequestHandler<LSP.DocumentOnAutoInsertPar
+            var newSourceText = sourceText.WithChanges(textChanges);
+
+            var caretLine = newSourceText.Lines[desiredCaretLinePosition.Line];
+            if (caretLine.IsEmptyOrWhitespace() && desiredCaretLinePosition.Character > caretLine.Span.Length)
```
i'm not opposed to this.  but part of me feels things would just be simpler if we only handled the 'empty' case.  handling the whitespace case makes it feel like you're likely writing code that we shouldn't be able to hit.

CyrusNajmabadi replied:
other things to be resilient for (or assert can't happen): what if it's a non-empty-non-whitespace line and the Character is still past the end?

i think it would make sense to potentially clamp the Character to the line, unless the line is purely empty.  might be simpler and easier to undestand/maintain.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/OnAutoInsert/OnAutoInsertHandler.cs:
```diff
@@ -145,11 +163,28 @@ internal class OnAutoInsertHandler : IRequestHandler<LSP.DocumentOnAutoInsertPar
+                var documentOptions = await originalDocument.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var indentText = amountToIndent.CreateIndentationString(documentOptions.GetOption(FormattingOptions.UseTabs), documentOptions.GetOption(FormattingOptions.TabSize));
+                var indentedText = textToIndent.WithChanges(new TextChange(new TextSpan(locationToIndent, 0), indentText));
+                return indentedText;
```
do you hae tests that test tabs?  we should def have those if not.

dibarbet replied:
good call, this uncovered a small bug I needed to fix :)

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractBraceCompletionServiceFactory.cs:
```diff
@@ -0,0 +1,38 @@
+
+        private static bool IsServiceValidForBraceCompletion(IBraceCompletionService service, Document document, int openingPosition, char openingBrace, CancellationToken cancellationToken)
+        {
+            return service.IsValidForBraceCompletionAsync(openingBrace, openingPosition, document, cancellationToken).WaitAndGetResult(cancellationToken);
```
ick that this is async here.  as elsewhere, i would prefer this moved up as much as possible

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -100,7 +101,14 @@ private void Start(CancellationToken cancellationToken)
+                var braceResult = _service.GetBraceCompletionAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
                 if (braceResult == null)
                 {
                     EndSession();
```
so ths pattern of having multiple locations need to 'EndSession' is error prone.  i.e. you could have easily added:

```
                if (context == null)
                {
                    return;
                }
```

and we likely wouldn't have noticed a bug. 

a nicer pattern i've found is more like:

```
if (!DoTheWork(args))
    EndSession();

//...

bool DoTheWork(... parameters ....)
{
    if (context == null)
        return false;
}

```

The benefit here is that the callee is forced to return a value, so they have to think about success/failure.  and tehn all failure is handle din a uniform fashion.



And

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -115,7 +123,14 @@ private void Start(CancellationToken cancellationToken)
 
-                var changesAfterStart = _session.GetChangesAfterCompletion(this, cancellationToken);
+                var contextAfterStart = GetBraceCompletionContext();
+                if (context == null)
```
wat?

context was already checked for null above.

dibarbet replied:
🤦 wrong context

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -115,7 +123,14 @@ private void Start(CancellationToken cancellationToken)
+                {
+                    caretPreservingTransaction.Complete();
+                    return;
+                }
```
this flow is very weird.  i would prefer a single place where we complete the transaction (since all paths complete here).  but have the special cases code if you `ApplyBraceCompletionResult`

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -217,6 +233,12 @@ public void PreOverType(out bool handledCommand)
                         }
                     }
                 }
+
```
`return;`

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -256,7 +278,13 @@ public void PostReturn()
+                            return;
+                        }
+
+                        var changesAfterReturn = _service.GetTextChangeAfterReturnAsync(context, CancellationToken.None).WaitAndGetResult(CancellationToken.None);
```
:'(  with all the CT.Nones.  We might want to file feedback with editor that these are potentially expensive and CT should be hooked up.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -112,20 +112,31 @@ protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken to
+                var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                return !syntaxFactsService.IsInNonUserCode(tree, currentPosition, cancellationToken);
             }
```
fwiw, this local function doesn't seem to be important :)  could just inline right into the outer method.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -112,20 +112,31 @@ protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken to
+        protected Task<bool> AllowOverTypeWithValidClosingTokenAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            return CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken);
         }
```
these helpers made it a lot eaier to read.  thansk! :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -144,21 +164,14 @@ public override async Task<bool> IsValidForBraceCompletionAsync(char brace, int
+            // Set the start point to the character after the opening brace.
+            var start = openingPosition + 1;
+            // Set the end point to the closing brace start character position.
+            var end = closingBraceEndPoint - 1;
```
👍

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/InterpolationBraceCompletionService.cs:
```diff
@@ -80,18 +89,8 @@ public static async Task<bool> IsPositionInInterpolationContextAsync(Document do
-                SyntaxKind.InterpolatedStringTextToken,
-                SyntaxKind.InterpolatedStringEndToken);
+            // We can be starting an interpolation expression if we're inside an interpolated string.
+            return token.GetAncestor((node) => node.IsKind(SyntaxKind.InterpolatedStringExpression)) != null;
```
hrmm... this is looking up arbitrarily far.  i.e. you could be in block in a lambda in an interpolation.  Is that ok?

dibarbet replied:
Ah good point...
I'll add a test case for that and see if I can work out a better check.  The reason I went for ancestor was that the InterpolatedStringExpression node could be a couple nodes up from the current token, but I definitely need to limit it

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -18,7 +18,15 @@ internal interface IBraceCompletionService : ILanguageService
+        /// The opening brace character to be inserted at the opening position.</param>
+        /// <param name="openingPosition">
+        /// The opening position to insert the brace.
+        /// Note that the brace is not yet inserted at this position in the document.
```
👍

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -18,7 +18,15 @@ internal interface IBraceCompletionService : ILanguageService
+        /// Note that the brace is not yet inserted at this position in the document.
+        /// </param>
+        /// <param name="document">The document to insert the brace at the position.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
```
feel free to not include this. (doesn't hurt though).

CyrusNajmabadi commented:
Overall looking great.  all the last commits made thingsa lot better.  Some minor cleanup requests (and a couple of potential WATs going on).  Will speak to you to get a sense of where we're landing on textchange merging.

dibarbet commented:
@CyrusNajmabadi I think you were right in that merging the formatting changes into the new line changes is easier than the general case.  I implemented the logic manually in https://github.com/dotnet/roslyn/pull/48468/commits/f770c7c61f0241b5a45aece1005f141b653dd348

ajaybhargavb commented:
@dibarbet, just as an FYI, I've merged the Razor side of the changes https://github.com/dotnet/aspnetcore-tooling/pull/2769. Things should just work with your changes after the next Razor insertion.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -64,23 +73,21 @@ private class BraceCompletionSession : IBraceCompletionSession
             }
 
-            private void Start(CancellationToken cancellationToken)
+            private bool Start(CancellationToken cancellationToken)
```
maybe call TryStart

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -64,23 +73,21 @@ private class BraceCompletionSession : IBraceCompletionSession
             public void Start()
             {
                 // Brace completion is not cancellable.
```
consider asserting you are on the Ui therad.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs:
```diff
@@ -262,13 +270,26 @@ public void PreReturn(out bool handledCommand)
+                            return;
+                        }
+
+                        var changesAfterReturn = _service.GetTextChangeAfterReturnAsync(context.Value, CancellationToken.None).WaitAndGetResult(CancellationToken.None);
```
i don't love hte inconsistency with how CancellationToken.NOne is treated across all these methods.

CyrusNajmabadi commented on src/EditorFeatures/VisualBasic/AutomaticCompletion/VisualBasicBraceCompletionServiceFactory.vb:
```diff
@@ -0,0 +1,24 @@
+        <ImportingConstructor>
+        <Obsolete(MefConstruction.ImportingConstructorMessage, True)>
+        Public Sub New(
+        <ImportMany(LanguageNames.VisualBasic)> braceCompletionServices As IEnumerable(Of IBraceCompletionService),
+        threadingContext As IThreadingContext)
```
```suggestion
                <ImportMany(LanguageNames.VisualBasic)> braceCompletionServices As IEnumerable(Of IBraceCompletionService),
                threadingContext As IThreadingContext)
```

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,511 @@
+        /// Annotation used to find the closing brace location after formatting changes are applied.
+        /// The closing brace location is then used as the caret location.
+        /// </summary>
+        private static readonly SyntaxAnnotation s_closingBraceSyntaxAnnotation = new("original closing brace");
```
```suggestion
        private static readonly SyntaxAnnotation s_closingBraceSyntaxAnnotation = new(nameof(s_closingBraceSyntaxAnnotation));
```

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,511 @@
+            {
+                newLineEdit = new TextChange(new TextSpan(closingPoint - 1, 0), Environment.NewLine);
+                textToFormat = originalDocumentText.WithChanges(newLineEdit.Value);
+            }
```
is this codepath effectively tested (both when it is/isn't hit)?

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,511 @@
+            }
+
+            // Modify the closing point location to adjust for the newly inserted line.
+            closingPoint += Environment.NewLine.Length;
```
don't love this.  it seems like it would actually be wrong in many circumstances (for example .editorconfig setting newline to '\n').  deserves tests of that :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,511 @@
+            closingPoint += Environment.NewLine.Length;
+
+            // Format the text that contains the newly inserted line.
+            var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
```
since you can get optins, we can definitely determine the real newline char.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,511 @@
+                return indentedLinePosition;
+            }
+
+            static ImmutableArray<TextChange> MergeFormatChangesIntoNewLineChange(TextChange newLineEdit, ImmutableArray<TextChange> formattingEdits)
```
i don't mind this.  but do we have appropriate testing of all the codepaths in this?

also, why was this better thna just exposing the existing merging function and just having us merge two lists togetehr where one list contains the single text change?

dibarbet replied:
Yeah I went this way for two main reasons
1.   Adding IVT from the compiler layer to the features layer here so I could use this caused a bunch errors across the sln due to types (like ArrayBuilder) that are part of both MS.CA.Workspaces and MS.CA.  When I have IVTs to both everything breaks `The type 'ArrayBuilder<T>' exists in both 'Microsoft.CodeAnalysis.Workspaces, Version=42.42.42.42, Culture=neutral, PublicKeyToken=31bf3856ad364e35' and 'Microsoft.CodeAnalysis, Version=42.42.42.42, Culture=neutral, PublicKeyToken=31bf3856ad364e35
`.  Not certain if there is a way to fix this, but making the whole type public didn't seem right.
2.  Based on your suggestion, this wasn't too complicated to write.

Will add tests

CyrusNajmabadi replied:
my *strong* preference would be to extract the algorithm into a static class/helper that is then just linked in (like we do with ArrayBuilder itself).  I would not expect IVT, just a sharing of the code here.

dibarbet replied:
Ah OK, sounds good I'll do that.

dibarbet replied:
@CyrusNajmabadi do you know of a good existing project for shared items like this?  It looks like MS.CA is using
Microsoft.CodeAnalysis.PooledObjects.projitems
Microsoft.CodeAnalysis.Debugging.projitems
AnalyzerDriver.projitems

But none of those seemed quite correct, and I didn't see any common project that is shared.  I can create a new projectitem for this but wanted to check if I was missing one.

CyrusNajmabadi replied:
I can look later. However, wherever the pile objects are shared seems like the right place

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,511 @@
+            return await base.CanProvideBraceCompletionAsync(brace, openingPosition, document, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected override bool IsValidOpeningBraceToken(SyntaxToken token) => token.IsKind(SyntaxKind.OpenBraceToken) && !token.Parent.IsKind(SyntaxKind.Interpolation);
```
```suggestion
        protected override bool IsValidOpeningBraceToken(SyntaxToken token)
            => token.IsKind(SyntaxKind.OpenBraceToken) && !token.Parent.IsKind(SyntaxKind.Interpolation);
```

so, for *methods*, we prefer to always wrap and indent for expr-bodies.  ihaven't minded up till now because teh methods were trivial.  but if there's going to be complexity, better to just wrap/indent all of them.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,511 @@
+                    startPoint--;
+                }
+
+                // skip token
```
```suggestion
                // skip open brace
```
?

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,511 @@
+        private static ImmutableArray<AbstractFormattingRule> GetBraceIndentationFormattingRules(DocumentOptionSet documentOptions)
+        {
+            var indentStyle = documentOptions.GetOption(FormattingOptions.SmartIndent);
+            return ImmutableArray.Create(BraceCompletionFormattingRule.ForIndentStyle(indentStyle));
```
do we adequately test the three different smart indent styles?

dibarbet replied:
Yeah we have tests for the cases (see IndentStyle).  If not specified it's the default value (smart)
http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests/AutomaticCompletion/AutomaticBraceCompletionTests.cs,1144

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/InterpolatedStringBraceCompletionService.cs:
```diff
@@ -0,0 +1,94 @@
+
+        protected override Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        {
+            return Task.FromResult(IsValidOpeningBraceToken(token)
+                && token.Span.End - 1 == position);
```
this read very weird.  consider either:

```suggestion
            return Task.FromResult(IsValidOpeningBraceToken(token) && token.Span.End - 1 == position);
```

or

```suggestion
            return Task.FromResult(
                IsValidOpeningBraceToken(token) && token.Span.End - 1 == position);
```

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/ParenthesisBraceCompletionService.cs:
```diff
@@ -0,0 +1,63 @@
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            if (ParentIsSkippedTokensTrivia(syntaxFactsService, token) ||
+                !IsValidOpeningBraceToken(token) ||
+                token.SpanStart != position || token.Parent == null)
```
```suggestion
                token.SpanStart != position ||
                token.Parent == null)
```

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/ParenthesisBraceCompletionService.cs:
```diff
@@ -0,0 +1,63 @@
+            }
+
+            // now check whether parser think whether there is already counterpart closing parenthesis
+            var (openBrace, closeBrace) = token.Parent.GetParentheses();
```
i don't understand this comment.  there is no check being done :D

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/ParenthesisBraceCompletionService.cs:
```diff
@@ -0,0 +1,63 @@
+            }
+
+            // We can complete the brace if the closing brace is missing or the incorrect kind.
+            return closeBrace.Kind() != SyntaxKind.CloseParenToken || closeBrace.Span.Length == 0;
```
oh.  i guess the check is here.  seems like thsi should be done first, before the more expensive checks :)

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,211 @@
+        /// </summary>
+        protected abstract bool IsValidClosingBraceToken(SyntaxToken token);
+
+        public abstract Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);
```
definitely doc.

dibarbet replied:
I had inheritdoc on this (since it's coming from the interface), but removed in response to https://github.com/dotnet/roslyn/pull/48468#discussion_r526236784

I'll leave this as-is since there are docs on the interface method already.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,211 @@
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return OpeningBrace == brace && !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
```
```suggestion
            return OpeningBrace == brace && !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
```

do this check first.  before the expensive work to get the tree.  this applies everywhere.  we do this work on every keystroke, so we want to bail out in the common cases as quickly as possibley.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,211 @@
+            var tree = await context.Document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = context.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return !syntaxFactsService.IsInNonUserCode(tree, context.CaretLocation, cancellationToken)
```
would it make sense at all to push the:

1. am i on a brace?
2. am i in non-user code?

all the way up to the top, so that this check is only in one plkace?

dibarbet replied:
These checks have to happen here because there are two entry points (which are triggered by two different locations in the users code for two different use cases).  They're basically at the 'top' already.
1.  Determining if this brace completion service can be used to complete a brace.  This also checks the opening _character_ (not inserted).
2.  Determining if overtyping is allowed at a location (and the non-user code check only applies to certain kinds of braces).  This also checks the closing _token_ kind (inserted)

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,211 @@
+            public const char CloseCharacter = '\'';
+        }
+
+        protected static async Task<bool> CouldEscapePreviousOpenBraceAsync(char openingBrace, int position, Document document, CancellationToken cancellationToken)
```
doc this.  i have no idea whatr's going on :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -102,10 +102,10 @@ public override Task<bool> AllowOverTypeAsync(BraceCompletionContext context, Ca
-            // Modify the closing point location to adjust for the newly inserted line.
-            closingPoint += Environment.NewLine.Length;
+                // Modify the closing point location to adjust for the newly inserted line.
+                closingPoint += Environment.NewLine.Length;
```
still not happy about using Env.NewLine.

dibarbet replied:
this is using doc options in a following commit https://github.com/dotnet/roslyn/pull/48468/commits/d9a327c9019b3e371f3a605dd7a0b5a261290419

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticBraceCompletionTests.cs:
```diff
@@ -1348,150 +1348,6 @@ public void CurlyBraceFormatting_InsertsCorrectNewLine()
-            Assert.Equal(new TextChange(new TextSpan(2, 1), "ee"), mergedChanges[3]);
-            Assert.Equal(new TextChange(new TextSpan(3, 1), "ff"), mergedChanges[4]);
-            Assert.Equal(incrementalText.ToString(), mergedText.ToString());
-        }
```
consider still having these tests, but potentially move to compiler layer.

dibarbet replied:
There's already a bunch of tests for the merge function here - http://sourceroslyn.io/#Microsoft.CodeAnalysis.UnitTests/Text/TextChangeTests.cs,20 - these end up just being dupes.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -163,6 +163,28 @@ static LinePosition GetIndentedLinePosition(Document document, SourceText source
+            {
+                var newRanges = TextChangeRangeExtensions.Merge(
+                    ImmutableArray.Create(newLineEdit.ToTextChangeRange()),
+                    formattingChanges.Select(f => f.ToTextChangeRange()).ToImmutableArray());
```
SelectAsArray

CyrusNajmabadi replied:
also awesome that we can now use this :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,461 @@
+                var openingPointLine = sourceText.Lines.GetLineFromPosition(openingPoint).LineNumber;
+                var closingPointLine = sourceText.Lines.GetLineFromPosition(closingPoint).LineNumber;
+
+                return closingPointLine - 1 > openingPointLine && sourceText.Lines[closingPointLine - 1].IsEmptyOrWhitespace();
```
should we be checking that there is *exactly* one empty line between the braces?  currently, this allows for many lines, where only the one before the close brace is empty.

CyrusNajmabadi replied:
could use GetLineBetweenCurlys here

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,461 @@
+                    var newTextChangeSpan = newRange.Span;
+                    // Get the text to put in the text change by looking at the span in the formatted text.
+                    // As the new range start is relative to the original text, we need to adjust it assuming the previous changes were applied
+                    // to get the correct start location in the formatted text.
```
not sure i understand.  but i trust you .  consider an exmaple.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,461 @@
+        /// </summary>
+        private static async Task<(ImmutableArray<TextChange> textChanges, int finalCurlyBraceEnd)> FormatTrackingSpanAsync(
+            Document document,
+            DocumentOptionSet documentOptions,
```
redundant given that we're passing Doucment?

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,461 @@
+            var newClosingPoint = newRoot.GetAnnotatedTokens(s_closingBraceSyntaxAnnotation).Single().SpanStart + 1;
+
+            var textChanges = result.GetTextChanges(cancellationToken).ToImmutableArray();
+            return (textChanges, newClosingPoint);
```
main question: is this tested well?  i don't understand a lick of it.  but if we have good test coverage (and this is mostly moved code) then i'm ok with that.

dibarbet replied:
Yeah, we have quite a few unit tests and integration tests for all the brace completion scenarios, so I'm relatively confident in these changes

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:
```diff
@@ -0,0 +1,461 @@
+            {
+                var originalRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var closeBraceToken = originalRoot.FindToken(closingBraceEndPoint - 1);
+                var newCloseBraceToken = closeBraceToken.WithAdditionalAnnotations(s_closingBraceSyntaxAnnotation);
```
1. would it be worth putting some asserts hwere that this found an actual code brace as expected?
2. we don't dive into trivia.  is that expected?  is brace completion only outside of trivia?

dibarbet replied:
1.  Sure!
2.  the curly brace completion service is only used outside of trivia - curly braces inside interpolated strings use the interpolation service, otherwise there isn't curly brace completion.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/InterpolationBraceCompletionService.cs:
```diff
@@ -0,0 +1,80 @@
+
+            // We can be starting an interpolation expression if we're inside an interpolated string.
+            return token.Parent.IsKind(SyntaxKind.InterpolatedStringExpression) || token.Parent.IsParentKind(SyntaxKind.InterpolatedStringExpression);
+        }
```
as before, i don't understand most of this.  so primary question is if this is tested well.  if so, i'm ok.

CyrusNajmabadi commented on src/Features/CSharp/Portable/BraceCompletion/ParenthesisBraceCompletionService.cs:
```diff
@@ -0,0 +1,64 @@
+            // brace completion session higher up on the stack.  If that's the case then we can
+            // complete the opening brace here, so return this as valid for completion.
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            return text.Lines.GetLineFromPosition(openBrace.SpanStart) == text.Lines.GetLineFromPosition(closeBrace.Span.End);
```
minor nit: i don't know i we guarantee object equality here for SoureceTextLines.  nbd, but i prefer doing `.LineNumber == ... .LineNumber`

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,219 @@
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        public virtual Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
```
hrmm... why are thse both virtual?  woudl it be better to make each subclass have to think about this?  or is the general idea that each subclass implements at least one of these?

dibarbet replied:
In almost all cases the sub classes don't override this - they don't do anything special.  The only one that does is curly brace completion

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,219 @@
+            }
+
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            return Task.FromResult(IsValidOpeningBraceToken(token) && !ParentIsSkippedTokensTrivia(syntaxFactsService, token));
```
aside: is it possible for any of these virtual methods to become sync?  or is there always at elast one case where we do something async in one of hte sublasses?

dibarbet replied:
There's a few cases in the implementations where we do async work :(

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs:
```diff
@@ -0,0 +1,219 @@
+        }
+
+        protected static bool ParentIsSkippedTokensTrivia(ISyntaxFactsService syntaxFactsService, SyntaxToken token)
+            => token.Parent == null || syntaxFactsService.IsSkippedTokensTrivia(token.Parent);
```
method body doesn't match method name.  not sure how relevant that is.  is this copied from previous code?

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,105 @@
+        /// Checks if this brace completion service should be the one used to provide brace completions at
+        /// the specified position with the specified opening brace.
+        /// Note that this does not mean that the service will provide brace completion, but that
+        /// only this service will be asked to provide them.
```
not exactly ertain what this means.

CyrusNajmabadi commented on src/Features/Core/Portable/BraceCompletion/IBraceCompletionService.cs:
```diff
@@ -0,0 +1,105 @@
+        /// <summary>
+        /// The caret location in the new text created by applying all <see cref="TextChanges"/>
+        /// to the input text.  Note the column in the line position can be virtual in that it points
+        /// to a location in the line which does not actually contain whitespace.
```
"Hosts can determine how best to handle that virtual location.  For example, placing the character in virtual space (when suppported), or inserting an appropriate number of spaces into the document".

CyrusNajmabadi commented on src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs:
```diff
@@ -154,5 +155,21 @@ public static async Task<NamingRule> GetApplicableNamingRuleAsync(this Document
             throw ExceptionUtilities.Unreachable;
         }
+
+        public static ImmutableArray<AbstractFormattingRule> GetFormattingRules(this Document document, IEnumerable<AbstractFormattingRule>? additionalRules, TextSpan span)
```
is this called from multiple places?

CyrusNajmabadi replied:
argument order is a bit odd.  i would expect it to be the required 'TextSpan' first, then the optional additionalRules next.

CyrusNajmabadi reviewed (approved):
LGTM.  handle nits as you see fit.

please confirm appropriate testing of some of hte hairier sections of code.  if not covered well, please beef things up :)

dibarbet commented:
@dotnet/roslyn-compiler / @RikkiGibson  Need a review on the compiler changes in https://github.com/dotnet/roslyn/pull/48468/commits/442970196d7daa7ca418e676f1f82e8f4caafe6d (should include all compiler changes).

This extracts out the textchangerange merging logic into a shared helper.  Normally the merging of text changes between versions is handled internally in SourceText (via ChangedText).  However, the documents in the IDE are based on [ChangedSourceText](http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.Text/Extensions.SnapshotSourceText.cs,275) which is a custom implementation that uses VS editor snapshots to merge text changes between versions.  Unfortunately I need to merge a set of text changes originally created from a VS editor snapshot, but the intermediate source text versions do not have corresponding VS editor snapshots.  So ChangedSourceText is not able to accurately compute the merged text changes and always returns the whole file as the diff.

To resolve this, I extracted out the merge logic from ChangedText so that I could manually merge the text changes between versions myself.  I did not want to copy the VS snapshot text into a new SourceText either as that could be expensive.

RikkiGibson commented on src/Compilers/Core/Portable/InternalUtilities/TextChangeRangeExtensions.cs:
```diff
@@ -94,5 +98,339 @@ internal static class TextChangeRangeExtensions
+        /// <remarks>
+        /// Both `oldChanges` and `newChanges` must contain non-overlapping spans in ascending order.
+        /// </remarks>
+        public static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRange> oldChanges, ImmutableArray<TextChangeRange> newChanges)
```
Consider moving back to the original location and just changing the accessibility.

dibarbet replied:
There are not any IVTs to Features.CSharp from MS.CA, and adding IVTs causes tons of build errors as now it thinks there are duplicate types (e.g. ArrayBuilder, DeclarationInfo, nullable attributes).

dibarbet commented:
@dotnet/roslyn-compiler could use one more review (thanks Rikki!).  Context is here - https://github.com/dotnet/roslyn/pull/48468#issuecomment-739020738

333fred reviewed (approved):
Compiler changes lgtm (commit 37)
